codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/ServerHttpSensitivityGenerator.kt

```diff
@@ -18,334 +15,435 @@ import software.amazon.smithy.model.traits.HttpLabelTrait
+    }
+
+    /** Returns the setter enclosing the closure or suffix position. */
+    fun setter(): Writable = if (hasRedactions()) writable {
+        rustTemplate(".label(#{Closure:W}, #{GreedyLabel:W})", "Closure" to closure(), "GreedyLabel" to greedyLabelStruct())
+    } else writable { }
+}
 
-        return if (valuesSensitive) {
-            // All values are sensitive
-            HeaderSensitivity.SensitiveMapValue(headerKeys, keySensitive, httpPrefixName)
+/** Models the ways headers can be bound and sensitive */
+sealed class HeaderSensitivity(
+    /** The values of the sensitive `httpHeaders`. */
+    val headerKeys: List<String>,
+    runtimeConfig: RuntimeConfig,
+) {
+    private val codegenScope = arrayOf(
+        "SmithyHttpServer" to ServerCargoDependency.SmithyHttpServer(runtimeConfig).asType(),
+        "Http" to CargoDependency.Http.asType(),
+    )
+
+    /** The case where `prefixHeaders` value is not sensitive. */
+    class NotSensitiveMapValue(
+        headerKeys: List<String>,
+        /** The value of `prefixHeaders`, null if it's not sensitive. */
+        val prefixHeader: String?,
+        runtimeConfig: RuntimeConfig,
+    ) : HeaderSensitivity(headerKeys, runtimeConfig)
+
+    /** The case where `prefixHeaders` value is sensitive. */
+    class SensitiveMapValue(
+        headerKeys: List<String>,
+        /** Is the `prefixHeaders` key sensitive? */
+        val keySensitive: Boolean,
+        /** The value of `prefixHeaders`. */
+        val prefixHeader: String,
+        runtimeConfig: RuntimeConfig,
+    ) : HeaderSensitivity(headerKeys, runtimeConfig)
+
+    /** Is there anything to redact? */
+    internal fun hasRedactions(): Boolean = headerKeys.isNotEmpty() || when (this) {
+        is NotSensitiveMapValue -> prefixHeader != null
+        is SensitiveMapValue -> true
+    }
+
+    /** Returns the type of the `MakeDebug`. */
+    fun type(): Writable = writable {
+        if (hasRedactions()) {
+            rustTemplate("#{SmithyHttpServer}::logging::sensitivity::headers::MakeHeaders<fn(&#{Http}::header::HeaderName) -> #{SmithyHttpServer}::logging::sensitivity::headers::HeaderMarker>", *codegenScope)
         } else {
-            HeaderSensitivity.NotSensitiveMapValue(headerKeys, httpPrefixName)
+            rustTemplate("#{SmithyHttpServer}::logging::MakeIdentity", *codegenScope)
         }
     }
 
-    internal fun renderHeaderClosure(writer: RustWriter, headerSensitivity: HeaderSensitivity) {
-        writer.rustBlockTemplate("|name: &#{Http}::header::HeaderName|", *codegenScope) {
-            rust("##[allow(unused_variables)]")
-            rust("let name = name.as_str();")
+    /** Returns the closure used during construction. */
+    internal fun closure(): Writable {
+        val nameMatch = if (headerKeys.isEmpty()) writable {
+            rust("false")
+        } else writable {
+            val matches = headerKeys.joinToString("|") { it.dq() }
+            rust("matches!(name, $matches)")
```

Comment:
Is `name` always used as `let name = name.as_str();`? If so, you could move `.as_str()` in `matches` and avoid the `allow_unused` (it's being added also when not needed)

Reply:
I think the closure might complain that `name` is unused entirely in this case.

Reply:
Worked, thanks! https://github.com/awslabs/smithy-rs/pull/1755/commits/b46b63e6894de5d07f7ede504048a7c2b1f81964